pkg/storage/utils/decomposedfs/migrator/migrator.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2018-2021 CERN
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package migrator
+
+import (
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"reflect"
+
+	"github.com/cs3org/reva/v2/pkg/storage/utils/decomposedfs/lookup"
+	"github.com/rogpeppe/go-internal/lockedfile"
+	"github.com/rs/zerolog"
+)
+
+var allMigrations = []string{"0001", "0002", "0003"}
+
+const (
+	resultFailed            = "failed"
+	resultSucceeded         = "succeeded"
+	resultSucceededRunAgain = "runagain"
+)
+
+type migrationState struct {
+	State   string
+	Message string
+}
+type migrationStates map[string]migrationState
+
+// Result represents the result of a migration run
+type Result string
+
+// Migrator runs migrations on an existing decomposedfs
+type Migrator struct {
+	lu     *lookup.Lookup
+	states migrationStates
+	log    *zerolog.Logger
+}
+
+// New returns a new Migrator instance
+func New(lu *lookup.Lookup, log *zerolog.Logger) Migrator {
+	return Migrator{
+		lu:  lu,
+		log: log,
+	}
+}
+
+// RunMigrations runs all migrations in sequence. Note this sequence must not be changed or it might
+// damage existing decomposed fs.
+func (m *Migrator) RunMigrations() error {
+	lock, err := lockedfile.OpenFile(filepath.Join(m.lu.InternalRoot(), ".migrations.lock"), os.O_WRONLY|os.O_CREATE, 0600)
+	if err != nil {
+		return err
+	}
+	defer lock.Close()
+
+	err = m.readStates()
+	if err != nil {
+		return err
+	}
+
+	for _, migration := range allMigrations {
+		s := m.states[migration]
+		switch s.State {
+		case "succeeded":
+			m.log.Info().Msg("Skipping finished migration " + migration + "...")
+			continue
+		case "":
+			m.log.Info().Msg("Running migration " + migration + "...")
+		default:
+			m.log.Info().Msg("Re-running migration " + migration + "...")
```

Comment:
we should probably drop any logs that show up during startup unless a migration actually needs to happen. We can think of passing down a logger or better just context the whole chain of services to clean this up in a subsequent PR.

But for now lets silence these logs, ok?

Reply:
Sure, I'll push an according commit soon.